### DISCUSSION TITLE
Why is vsg::Command::record() const?

### DISCUSSION THREAD

**Discussion by dov** [vsg-dev/VulkanSceneGraph#1294] [General]
I've been working on integrating ImGuizmo with vsg, and in particular, porting https://github.com/CedricGuillemet/ImGuizmo/blob/master/example/main.cpp . One thing that I stumbled on, was the integration of the `TransformStart()` and `TransformEdit()` functions in the above `main.cpp` into a `vsg::Command` class record function. After some struggling, I managed to port the code so that they are called from `vsg::Command::record`. This worked when keeping the functions external to the class as in the original example. I then wanted to encapsulate the class, and bring everything into member functions that are called from the `vsg::Command::record()` method. However, I then I got stuck on the issue of `record()` being defined `const`. 

The `TransformStart()` and `TransformEdit()` change the gizmo state, and saves the result in global variables. However, when moving these global variables into class variables, I can no longer change them in a function called from `record()` because of the `const` declaration.

So I'm wondering what's the best way of solving this? Here are some options that I can think of:

- Add another state-changing method in parallel to `record()` and call it from the inner loop prior to calling `viewer->recordAndSubmit()`
- Just keep the variables defining the state as global (or call static) variables, and share the state among all "gizmos"
- Remove the `const` declaration from `vsg::Command::record()` to allow for more flexibility?

I'd be happy to receive some feedback, about what you think is the best approach.

**Comment by robertosfield** [comment]
Command::record() is const as the intent is that users will just dispatch data to Vulkan and not modify the local data, so can the commands can be safely recorded from multi-threads without having to protect the data with mutexes.

If you have a special case Command that needs to modify it's local data the either cast away constness locally or declare the member variables as mutable.  You will need take care over threading of this Command so you don't end end with race conditions.
